File: src/main/java/fr/ayfri/minecraft_art/DrawZone.java
package fr.ayfri.minecraft_art;

import processing.awt.PGraphicsJava2D;
import processing.core.PImage;

public class DrawZone extends PGraphicsJava2D {
	private final Main sketch;
	private final int x;
	private final int y;
	
	public DrawZone(Main sketch, int x, int y) {
		this(sketch, x, y, 512,512);
	}
	
	public DrawZone(Main sketch, int x, int y, int width, int length) {
		this.sketch = sketch;
		this.width = width;
		this.height = length;
		this.x = x;
		this.y = y;
		setParent(sketch);
		setPrimary(false);
		setSize(width, height);
	}
	
	public void draw(PImage renderer) {
		beginDraw();
		strokeWeight(4);
		background(0);
		image(renderer, 0, 0, width, height);
		endDraw();
		
		sketch.image(this, x, y);
	}
	
	public void show() {
		sketch.image(this, x, y);
	}
}

File: src/main/kotlin/io/github/ayfri/minecraft_art/rendering/DrawZone.kt
package io.github.ayfri.minecraft_art.rendering

import io.github.ayfri.minecraft_art.Main
import processing.awt.PGraphicsJava2D
import processing.core.PImage

open class DrawZone(
	protected val sketch: Main,
	protected val x: Int,
	protected val y: Int,
	width: Int = 512,
	height: Int = 512,
) : PGraphicsJava2D() {
	init {
		setParent(sketch)
		setPrimary(false)
		setSize(width, height)
		/** Blocks are 16x16 pixel art, they must stay crisp. Processing only accepts this before the first beginDraw(). */
		noSmooth()
	}

	fun draw(renderer: PImage) {
		beginDraw()
		strokeWeight(4f)
		background(0)
		image(renderer, 0f, 0f, width.toFloat(), height.toFloat())
		noStroke()
		endDraw()

		show()
	}

	/** Renders the empty placeholder used while no image is loaded yet. */
	fun clearZone() {
		beginDraw()
		background(0)
		endDraw()

		show()
	}

	fun show() = sketch.image(this, x.toFloat(), y.toFloat())
}
